### DISCUSSION TITLE
Does not match non-negated prhases, with "+", "?" or "!" operator as first token

### DISCUSSION THREAD

**Discussion by nomoteticus** [explosion/spaCy#5538] [Help: Coding & Implementations] [answered]
I am trying to match a short phrase where the token is not negated.
Fake - should match
Kinda fake - should match
Not fake - should not match
Not really fake - should not match

So I am trying to put the negation as an optional token at the beginning of the matcher.
I tried with the ? operator and + operator, also tried with the ! operator (with IN instead of NOT_IN). Nothing works

## How to reproduce the behaviour

pattern_o  = \
   [{"LEMMA":{"NOT_IN":['not','no']},"OP":"+"},
    {"DEP":"ROOT", 
     "LEMMA":{"IN":['fake','false']}}]

Result(for o1):
Fake - does not match (should match)
Kinda fake - matches (ok)
Not fake - does not match (ok)
Not really fake - matches (should not match)

pattern_o2  = \
   [{"LEMMA":{"NOT_IN":['not','no']},"OP":"?"},
    {"DEP":"ROOT", 
     "LEMMA":{"IN":['fake','false']}}]
pattern_o3  = \
   [{"LEMMA":{"IN":['not','no']},"OP":"!"},
    {"DEP":"ROOT", 
     "LEMMA":{"IN":['fake','false']}}]

Result (for o2 and o3):
Fake - matches (ok)
Kinda fake - matches (ok)
Not fake - matches (should not match)
Not really fake - matches (should not match)

## Your Environment

* Operating System: Ubuntu 18.10
* Python Version Used: 3.8.2
* spaCy Version Used: 2.2.4
* Environment Information:

**Comment by adrianeboyd** [comment, marked as answer]
I think the problem is some confusion about the definition of the quantifiers. `!` in particular can cause some confusion and I think the docs could be improved a bit here.

`{"OP": "!"}` matches exactly one negated token and won't match if no token is present in that position in the pattern. It can't be used to match 0 tokens.

`{"OP": "?"}` can match 0 or 1 tokens, so you're see some matches on part of the text, like just the word `fake` rather than the whole text. If you only want matches that cover the whole text, you'll have to filter the matches (the `on_match` callback is an easy place to do this).
